domain_test.go

```diff
@@ -61,6 +61,84 @@ func TestDomainGetSystems(t *testing.T) {
 	autopilot.Equals(t, "PlatformSystem3", result[2].Name)
 }
 
+func TestDomainReconcileAliasesDeleteAll(t *testing.T) {
+	// Arrange
+	aliasesWanted := []string{}
+	domain := ol.Domain{
+		DomainId: ol.DomainId{
+			Id: id1,
+		},
+		ManagedAliases: []string{"one", "two"},
+	}
+
+	// delete "one" alias
+	testRequestOne := autopilot.NewTestRequest(
```

Comment:
More detailed explanation of what I meant here after talking with @davidbloss 

1. `ExtractAliases()` is a really easily testable function because it returns the list of aliases that are to be deleted, and the list of aliases that are to be added.
2. `ReconcileAliases()` just does `ExtractAliases()` + `client.DeleteAlias()` and `client.AddAlias()` multiple times. This means that `ExtractAliases()` is kind of redundant and should at least be made private since it has no use outside of reconciliation.
3. `ReconcileAliases()` should be extremely explicit about what it did. It should not just return `errors` but also return a list of aliases that were added, aliases that were deleted. This way we know 100.0% that it is **making the correct decision to add and delete certain aliases** as part of reconciliation, and **we also know that the API calls it is making are valid API calls.**
4. Another way to solve this would be to add tooling to `BestTestClient` or the `autopilot` package that would let us setup a more explicit test expectation like:

```
client = newclient()
client.expects_calls = [{ full json for the first API call to be made }, { full json for the second API call to be made }]
domain.ReconcileAliases(client, wantedAliases)

# if client does not receive 2 calls, the test will fail
# if client does not receive call 1 and call 2 with exactly the expected json request defined in the test the test should fail.
```

^These types of test expectations can be set in Ruby test mocking libraries like `mocha`. We can do the same in Go but we need to write more code to support it.

---

I am lifting my 'request changes' for this because

1. I am certain the way the code is now that the API calls being made during reconciliation are correct. But the implementation is still subject to change.
2. There are MANY parts of the codebase where we don't have explicit expectations set like mentioned in (4).

Reply:
To add to (3), you can take this test and add on `testRequestThree` and `testRequestFour` and so on and it will still pass, which would be incorrect API requests for the reconcile function to make but the test would still pass because it is just checking that the result is correct.

Reply:
We can only confirm with `autopilot` that:

* That up to N API calls are made
* That if an API call is made, it is one that has been "registered" with autopilot.
* It does not account for less than N API calls being made or for the same API call being made several times. Or for an API call that was "registered" not being made. The test expectations outside of the API call logic have to deal with that.